[WIP] [First Draft] NE753: Docs work for External DNS Operator feature (#40441)
openshift/openshift-docs: networking/external_dns_operator/nw-external-dns-operator-configuration-parameters.adoc

[id="nw-external-dns-operator-configuration-parameters_{context}"]
= External DNS Operator configuration parameters

[cols="3a,8a",options="header"]
|===
|Parameter |Description

|`spec`
|`spec` parameter enables

[source,yaml]
----
spec:
  provider:
    type: AWS <1>
    aws:
      credentials:
        name: aws-access-key <2>
----
<1>
<2>

|`zones`

> |`zones` enables your specified AWS zone. When you do not specify zones, ExternalDNS discovers all the zones in the account.

Author: ack

|`domains`
|`domains` parameter manages hostnames.

[source,yaml]
----
domains:
- filterType: Include <1>
  matchType: Exact <2>
  name: "myzonedomain1.com" <3>
- filterType: Include
  matchType: Pattern <4>
  pattern: ".*\\.otherzonedomain.com" <5>
----
<1>
<2>
<3>
<4>
<5>

|`source`
|`source`

[source,yaml]
----
source:
  type: Service <1>
  service:
    serviceType:
      - LoadBalancer <2>
      - ClusterIP <3>
  labelFilter:

>     matchLabels:

Reviewer: Specifying a label filter ensures that the only those resources with the matching label filter will be considered by the controller. Resources with a matching label filter will have DNS records created for them based on the configuration.

>     matchLabels:

Reviewer 2: @xenolinux  ^^

>     matchLabels:

Reviewer: https://github.com/openshift/openshift-docs/pull/40441/files#r797548418

>     matchLabels:

Author: ack

      external-dns.mydomain.org/publish: "yes" <4>

>   hostnameAnnotation: "Allow" <5>

Reviewer 2: Allow hostname annotation when generating DNS names. @arjunrn  can you provide more details on hostname annotation as you have worked on this ?
Default value is Ignore and when it is Ignore one needs to set fqdnTemplate.

Reviewer 2: @arjunrn can you please check the above ?

Reviewer: https://github.com/openshift/openshift-docs/pull/40441/files#r797553673

Author: ack

  fqdnTemplate:

>   - "{{.Name}}.myzonedomain.com" <6>

Reviewer 2: A templated string that's used to generate DNS names from sources that don't define a hostname themselves, or to add a hostname suffix when paired with the fake source.

Author: ack

----

<1>
<2>
<3>
<4>
<5>
<6>

[source,yaml]
----
source:
  type: OpenShiftRoute <1>
  openshiftRouteOptions:
    routerName: default <2>
----

> <1>

Reviewer 2: You can define an option for route source under this field. External DNS will filter routes for creation of dns records in the cloud dns.

> <2>

Reviewer 2: If the source is openshift-route then you can pass the ingress controller name. Based on this name external-dns will select the respective router from the route status and map that routerCanonicalHostname to the route host while creating a CNAME record.

<2> If the source is route then the route hostname's domain will be checked for the name defined in the name field. So, as per this example if an hostname is test.myzonedomain1.com or apps.test.myzonedomain1.com then it will be filtered by external dns for dns record creation in the defined zone ID's present under zone's section.
<3> Define the domain name to be filtered.
<4> Sets "regex-domain-filter" in external-dns. Limit possible domains and target zones by a Regex filter.
<5> Define the regex pattern to be used by external to filter domains and target zones.

|`source`
|`source`

[source,yaml]
----
source: <1>
  type: Service <2>
  service:
    serviceType:<3>
      - LoadBalancer
      - ClusterIP
  labelFilter: <4>
    matchLabels:
      external-dns.mydomain.org/publish: "yes"
  hostnameAnnotation: "Allow" <5>
  fqdnTemplate:
  - "{{.Name}}.myzonedomain.com" <6>
----

<1> The resource types that are queried for endpoints.
<2> External DNS will filter source type of service for creating dns records.
<3> It sets "service-type-filter" in external-dns. The service types to take care about (default: all, expected: ClusterIP, NodePort, LoadBalancer or ExternalName)

> <4> Filter sources managed by external-dns via label selector when listing all resources; currently supported by source types CRD, ingress, service and openshift-route".

Reviewer: The description of `<4>` is incorrect. The `labelFilter` field is used to specify a [LabelSelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#labelselector-v1-meta). It can have `matchLabels` or `matchExpression` and this selector is used to filter out resources which don't have the expected labels.

> <5> Allow hostname annotation when generating DNS names. Default value is Ignore and when it is Ignore one needs to set fqdnTemplate.

Reviewer: When the value is `"Allow"` the DNS records will be generated based on the value specified in the `external-dns.alpha.kubernetes.io/hostname` specified on the resource. The default value is `"Ignore"` which means that the DNS record will be generated using the templates specified in the field `fqdnTemplates`. The default value is `"Ignore"`.

<6> A templated string that's used to generate DNS names from sources that don't define a hostname themselves, or to add a hostname suffix when paired with the fake source.

[source,yaml]
----
source:
  type: OpenShiftRoute <1>
  openshiftRouteOptions:
    routerName: default <2>
----

<1> You can define an option for route source under this field. External DNS will filter routes for creation of dns records in the cloud dns.

> <2> If the source is openshift-route then you can pass the ingress controller name. Based on this name external-dns will select the respective router from the route status and map that routerCanonicalHostname to the route host while creating a CNAME record.

Reviewer 3: ```suggestion
<2> If the source is openshift-route then you can pass the ingress controller name. Based on this name external-dns will select the respective router's canonical host name to be put as target while creating a CNAME record.
```
